File: solutions/go/pkg/year2022/day02.go
package year2022

import (
	"fmt"
	"strings"
)

const (
	Loss int = 0
	Draw int = 3
	Win  int = 6
)

const (
	Rock     int = 1
	Paper    int = 2
	Scissors int = 3
)

type Round struct {
	Player  int
	Oponent int
}

func parseOponentShape(encoded string) int {
	switch encoded {
	case "A":
		return Rock
	case "B":
		return Paper
	case "C":
		return Scissors
	default:
		panic(fmt.Errorf("Unexpected oponent shape: %s", encoded))
	}
}

func parsePlayerShape(encoded string) int {
	switch encoded {
	case "X":
		return Rock
	case "Y":
		return Paper
	case "Z":
		return Scissors
	default:
		panic(fmt.Errorf("Unexpected player shape: %s", encoded))
	}
}

func parsePlayerResult(encoded string) int {
	switch encoded {
	case "X":
		return Loss
	case "Y":
		return Draw
	case "Z":
		return Win
	default:
		panic(fmt.Errorf("Unexpected player result: %s", encoded))
	}
}

func calculateRoundResult(player, oponent int) int {
	if (player == Rock && oponent == Scissors) ||
		(player == Paper && oponent == Rock) ||
		(player == Scissors && oponent == Paper) {
		return Win
	}

	if (player == Scissors && oponent == Rock) ||
		(player == Paper && oponent == Scissors) ||
		(player == Rock && oponent == Paper) {
		return Loss
	}

	return Draw
}

func getNecessaryShapeForResult(result, oponent int) int {
	if (result == Loss && oponent == Paper) ||
		(result == Draw && oponent == Rock) ||
		(result == Win && oponent == Scissors) {
		return Rock
	}

	if (result == Loss && oponent == Scissors) ||
		(result == Draw && oponent == Paper) ||
		(result == Win && oponent == Rock) {
		return Paper
	}

	return Scissors
}

func parseGame(input string, parsePlayer func(string) int) []Round {
	var game []Round

	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		split := strings.Split(line, " ")
		oponent := parseOponentShape(split[0])
		player := parsePlayer(strings.TrimSpace(split[1]))
		game = append(game, Round{player, oponent})
	}

	return game
}

func playGame(game []Round, calculatePoints func(int, int) int) int {
	totalPoints := 0

	for _, round := range game {
		totalPoints += calculatePoints(round.Player, round.Oponent)
	}

	return totalPoints
}

func day2Part1(game []Round) int {
	return playGame(game, func(player, oponent int) int {
		return player + calculateRoundResult(player, oponent)
	})
}

func day2Part2(game []Round) int {
	return playGame(game, func(player, oponent int) int {
		return player + getNecessaryShapeForResult(player, oponent)
	})
}

func Day2(input string) (int, int) {
	part1 := parseGame(input, parsePlayerShape)
	part2 := parseGame(input, parsePlayerResult)

	return day2Part1(part1), day2Part2(part2)
}

File: solutions/haskell/src/Year2022/Day02.hs
module Year2022.Day02 (solution) where

data Shape = Rock | Paper | Scissors deriving (Show, Eq)

data RoundResult = Win | Draw | Loss deriving (Show, Eq)

shapePoints :: Shape -> Int
shapePoints shape = case shape of
    Rock -> 1
    Paper -> 2
    Scissors -> 3

roundPoints :: RoundResult -> Int
roundPoints result = case result of
    Win -> 6
    Draw -> 3
    Loss -> 0

parseOponent :: String -> Shape
parseOponent encoded = case encoded of
    "A" -> Rock
    "B" -> Paper
    "C" -> Scissors
    _ -> error ("Parse oponent shape error: " ++ encoded)

parsePlayerIntoShape :: String -> Shape
parsePlayerIntoShape encoded = case encoded of
    "X" -> Rock
    "Y" -> Paper
    "Z" -> Scissors
    _ -> error ("Parse player shape error" ++ encoded)

parsePlayerIntoRoundResult :: String -> RoundResult
parsePlayerIntoRoundResult encoded = case encoded of
    "X" -> Loss
    "Y" -> Draw
    "Z" -> Win
    _ -> error ("Parse player desired result error: " ++ encoded)

parse :: (String -> p) -> String -> [(Shape, p)]
parse parsePlayer input = map (parseLine . words) $ lines input
  where
    parseLine line = case line of
        [oponent, player] -> (parseOponent oponent, parsePlayer player)
        _ -> error ("Parse error at line: '" ++ unwords line ++ "'")

playeGame :: (Shape -> p -> Int) -> [(Shape, p)] -> Int
playeGame calculatePoints game = sum $ map (uncurry calculatePoints) game

roundResult :: Shape -> Shape -> RoundResult
roundResult oponent player
    | (player, oponent) `elem` win = Win
    | (player, oponent) `elem` loss = Loss
    | otherwise = Draw
  where
    win = [(Rock, Scissors), (Paper, Rock), (Scissors, Paper)]
    loss = [(Scissors, Rock), (Paper, Scissors), (Rock, Paper)]

necessaryShape :: Shape -> RoundResult -> Shape
necessaryShape oponent player
    | (player, oponent) `elem` rockNeeded = Rock
    | (player, oponent) `elem` paperNeeded = Paper
    | otherwise = Scissors
  where
    rockNeeded = [(Loss, Paper), (Draw, Rock), (Win, Scissors)]
    paperNeeded = [(Loss, Scissors), (Draw, Paper), (Win, Rock)]

part1 :: [(Shape, Shape)] -> Int
part1 = playeGame calculatePoints
  where
    calculatePoints oponent player = roundPoints (roundResult oponent player) + shapePoints player

part2 :: [(Shape, RoundResult)] -> Int
part2 = playeGame calculatePoints
  where
    calculatePoints oponent player = roundPoints player + shapePoints (necessaryShape oponent player)

solution :: String -> (String, String)
solution input =
    ( show $ part1 $ parse parsePlayerIntoShape input
    , show $ part2 $ parse parsePlayerIntoRoundResult input
    )
